File: 738-MonotoneIncreasingDigits/738-MonotoneIncreasingDigits.cpp
// Last updated: 12/8/2025, 11:27:50 PM
class Solution {
public:
    int monotoneIncreasingDigits(int n) {
        string s = to_string(n);
        
        int start = 0;  // if s[i] == s[i + 1] == s[i + 2] == ... , set start as the first.
        int end = -1;   // if s[i] > s[i + 1], ends at start.
        for (int i = 1; i < s.length(); ++i) {
            if (s[i - 1] == s[i]) {
                continue;
            } else if (s[i - 1] < s[i]) {
                start = i;
            } else {
                end = start;
                break;
            }
        }
        // if break, build the new integer
        if (end != -1) {
            int ans = 0;
            // normally build before the end;
            for (int i = 0; i < end; ++i) {
                ans = ans * 10 + (s[i] - '0');
            }
            // decreast s[end] by 1
            ans = ans * 10 + (s[end] - '0' - 1);
            end++;
            // set the remaining digits to 9
            while (end < s.length()) {
                ans = ans * 10 + 9;
                end++;
            }
            return ans;
        }
        // if not break, return original integer
        return n;
    }
};

File: 738-MonotoneIncreasingDigits/738-MonotoneIncreasingDigits.py
# Last updated: 12/8/2025, 11:27:51 PM
divs = [100000000, 10000000, 1000000, 100000, 10000, 1000, 100, 10, 1]
class Solution:
    def monotoneIncreasingDigits(self, N: int) -> int:
        preStratDiv = 1000000000
        pre = N // preStratDiv
        
        curDiv = preStratDiv // 10
        Nmod = N
        for curDiv in divs:
            cur, nxtNmod = divmod(Nmod, curDiv)
            if cur >= pre:
                if cur > pre:
                    pre = cur
                    preStratDiv = curDiv
            else:
                N = (N // preStratDiv) * preStratDiv - 1
                break

            Nmod = nxtNmod
        
        return N